quickwit-metastore/src/metastore/single_file_metastore.rs

```diff
@@ -0,0 +1,1352 @@
+/*
+    Quickwit
+    Copyright (C) 2021 Quickwit Inc.
+    Quickwit is offered under the AGPL v3.0 and as commercial software.
+    For commercial licensing, contact us at hello@quickwit.io.
+    AGPL:
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Affero General Public License as
+    published by the Free Software Foundation, either version 3 of the
+    License, or (at your option) any later version.
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use std::collections::HashMap;
+use std::ops::Range;
+use std::path::Path;
+use std::sync::{Arc, RwLock};
+
+use async_trait::async_trait;
+use quickwit_doc_mapping::DocMapping;
+use quickwit_storage::{PutPayload, Storage};
+
+use crate::{
+    IndexMetadata, IndexUri, MetadataSet, Metastore, MetastoreErrorKind, MetastoreResult, SplitId,
+    SplitMetadata, SplitState, FILE_FORMAT_VERSION,
+};
+
+/// Single file meta store implementation.
+pub struct SingleFileMetastore {
+    storage: Arc<dyn Storage>,
+    data: Arc<RwLock<HashMap<IndexUri, MetadataSet>>>,
```

Comment:
let's use a tokio RwLock in async functions. 

They are not blocking...

Reply:
Changed to use tokio RwLock.